api/processors/auxil/algorithms.py

```diff
@@ -0,0 +1,171 @@
+# DO NOT USE IN PROD, WORK IN PROGRESS
+
+from collections.abc import Iterable
+from dataclasses import dataclass
+from datetime import datetime
+
+from django.utils import timezone
+
+from api.models import AgeRange, Group, Student, Teacher
+from api.models.auxil.constants import (
+    DEFAULT_LESSON_DURATION_MIN,
+    MAX_AGE_KIDS_GROUP,
+    MAX_AGE_TEEN_GROUP,
+)
+from api.models.auxil.status_setter import StatusSetter
+from api.models.choices.communication_language_mode import CommunicationLanguageMode
+from api.models.choices.log_event_type import (
+    GroupLogEventType,
+    StudentLogEventType,
+    TeacherLogEventType,
+)
+from api.models.choices.status import (
+    GroupProjectStatus,
+    StudentSituationalStatus,
+    TeacherProjectStatus,
+    TeacherSituationalStatus,
+)
+from api.models.language_and_level import LanguageAndLevel
+from api.processors.auxil.log_event_creator import GroupLogEventCreator
+
+
+class GroupBuilderAlgorithm:
+    @dataclass
+    class GroupSizeRestriction:
+        # both min and max student numbers are inclusive:
+        # group size must fit into [min; max]
+        min: int
+        max: int
+
+        def validate_size(self, size: int) -> None:
+            if self.min > size:
+                raise ValueError(
+                    f"Group candidate must respect upper group boundaries, "
+                    f"but {size} > {self.max}"
+                )
+            if self.max < size:
+                raise ValueError(
+                    f"Group candidate must respect upper group boundaries, "
+                    f"but {size} > {self.max}"
+                )
+
+    @dataclass
+    class GroupCandidate:
+        language_and_level: LanguageAndLevel
+        communication_language_mode: CommunicationLanguageMode
+        age_range: AgeRange
+        teacher: Teacher
+        students: list[Student]
+        monday: datetime | None = None
+        tuesday: datetime | None = None
+        wednesday: datetime | None = None
+        thursday: datetime | None = None
+        friday: datetime | None = None
+        saturday: datetime | None = None
+        sunday: datetime | None = None
+
+        def __post_init__(self) -> None:
+            # Validate group size against age-related restrictions
+            restrictions = GroupBuilderAlgorithm._get_allowed_group_size(self.age_range)
+            restrictions.validate_size(len(self.students))
+
+    @staticmethod
+    def get_available_teachers() -> Iterable[Teacher]:
```

Comment:
Logic changed since the comment, now it's not even a QuerySet :)

Otherwise I feel like sometimes moving one level of abstraction up improves readability, if the actual return type is overly specific.